---
title: "Untitled"
output: 
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Libraries

```{r}
library(ggplot2)
library(dplyr)
library(ape)
library(msa) ## need XVector, Biostrings. All through biocLite
library(seqinr)
library(RColorBrewer)
library(reshape2)
```

Load data

```{r}
df = read.table("140117.cdr12.txt", header = T, sep = "\t", stringsAsFactors = F) %>%
  filter(species == "HomoSapiens", grepl("TR[AB].+\\*01", gene)) %>%
  dplyr::select(gene, seqaa, cdr1aa, cdr2aa) %>%
  mutate(chain = ifelse(grepl("TRA", gene), "TRA", "TRB"),
         gene = gsub("\\*01", "", gene))

df.prob = read.table("cdr.contact.txt", header = T, sep = "\t")
df.prob$mhc_type = "any"
df.prob = rbind(df.prob, read.table("cdr.mhc.contact.txt", header = T, sep = "\t"))
```

## Multiple alignment of amino acid sequences

Set up plotting

```{r}
df.trb = subset(df, chain == "TRB")

clrs = colorRampPalette(rev(brewer.pal(9, 'Paired')))(length(df.trb$gene))

seq = df.trb$seqaa
names(seq) = df.trb$gene

trbAln = msa(AAStringSet(seq))

print(trbAln)
  
names(clrs) = names(trbAln@unmasked)

plot_tree = function(seqs, tip_colors = clrs) {
  trbAln = msa(AAStringSet(seq))
  
  tip_colors = tip_colors[names(trbAln@unmasked)]

  trbAln2 = msaConvert(trbAln, type="seqinr::alignment")
  d = dist.alignment(trbAln2, "identity")
  trbTree = nj(d)

  fig = c(0.1, 0.9, 0, 1.0)
  mar = c(0, 0, 0, 0)
  par(fig = fig, mar = mar, xpd = NA)#, bg = "grey")
  plot(trbTree, type = "fan",  tip.color = tip_colors)
  
  trbTree
}
```

### Full V length

```{r}
seq = df.trb$seqaa
names(seq) = df.trb$gene
full_phylo = plot_tree(seq)
```

### CDR1

```{r}
seq = df.trb$cdr1aa
names(seq) = df.trb$gene
cdr1_phylo = plot_tree(seq)
```

### CDR2

```{r}
seq = df.trb$cdr2aa
names(seq) = df.trb$gene
cdr2_phylo = plot_tree(seq)
```


### CDR1+2

```{r}
seq = paste0(df.trb$cdr1aa, df.trb$cdr2aa)
names(seq) = df.trb$gene
cdr12_phylo = plot_tree(seq)
```

### Compare trees

```{r}
fig = c(0.1, 0.9, 0, 1.0)
mar = c(0, 0, 0, 0)
par(fig = fig, mar = mar, xpd = NA, cex=0.3)
```

Full V alignment vs CDR1

```{r}
cophyloplot(full_phylo, cdr1_phylo, assoc = cbind(df.trb$gene, df.trb$gene), space=200, col = "red")
```

Full V alignment vs CDR2

```{r}
cophyloplot(full_phylo, cdr2_phylo, assoc = cbind(df.trb$gene, df.trb$gene), space=200, col = "red")
```

Full V alignment vs CDR1+2

```{r}
cophyloplot(full_phylo, cdr12_phylo, assoc = cbind(df.trb$gene, df.trb$gene), space=200, col = "red")
```

## Computing number of possible antigen contacts

Compute contacts

```{r}
pos_max = df.prob$pos_max[1]

compute_contacts = function(tcr_chain, tcr_region, region_seq, mhc_type = "any") {
  df.seq = data.frame(mhc_type = mhc_type,
                      tcr_chain = tcr_chain,
                      tcr_region = tcr_region,
                      aa_tcr = strsplit(region_seq, "")[[1]])
  
  df.seq$pos_rel_tcr = with(df.seq, (1:length(aa_tcr) - 1) / (length(aa_tcr) - 1) * (pos_max - 1))

  df.seq = merge(df.seq, df.prob)
  
  sum(df.seq$P)
}

df.cont = df %>%
  group_by(gene) %>%
  mutate(P = compute_contacts(chain, "CDR1", cdr1aa) + compute_contacts(chain, "CDR2", cdr2aa))
```

```{r}
clrs2 = colorRampPalette(c("#2166ac", "#bababa", "#b2182b"))(101)

df.cont$Pl = rank(df.cont$P)
df.cont$Pind = with(df.cont, 100 * (Pl - min(Pl)) / (max(Pl) - min(Pl)) + 1)

clrs2 = clrs2[df.cont$Pind]
names(clrs2) = df.cont$gene
```

TCR beta chain V segment alignment colored by net contact probability, from blue (low) to red (high)

```{r}
seq = df.trb$seqaa
names(seq) = df.trb$gene
full_phylo = plot_tree(seq, clrs2)
```

TCR alpha chain V segment alignment colored by net contact probability, from blue (low) to red (high)

```{r}
df.tra = subset(df, chain == "TRA")
seq = df.tra$seqaa
names(seq) = df.tra$gene
full_phylo = plot_tree(seq, clrs2)
```

## Frequency in repertoire

We'll study V segment usage in naive compartment. There are two factors that influence V usage: assembly probability and thymic selection. The latter should be negatively correlated with the number of contacts.

```{r}
load_v_usage = function(f) {
  df.vusage = read.table(paste0(f, ".segments.unwt.V.txt"), header=T, sep="\t") %>%
    filter(type == "naive")

  df.vusage$replica = as.factor(df.vusage$replica)

  df.vusage = melt(df.vusage)

  df.vusage$variable = gsub("\\.D", "/D", df.vusage$variable)
  df.vusage$variable = gsub("\\.", "-", df.vusage$variable)

  df.vusage = df.vusage %>% 
    mutate(chain2 = ifelse(grepl("TRA", variable), "alpha", "beta")) %>%
    filter(chain == chain2)
  
  as.data.frame(df.vusage)
}

df.vusage.f = load_v_usage("func")
df.vusage.nf = load_v_usage("nonfunc")

df.vusage = merge(df.vusage.f, df.vusage.nf, by = colnames(df.vusage.f)[!(colnames(df.vusage.f) %in% c("value", "..filter.."))])
df.vusage$freq.ratio = df.vusage$value.x / df.vusage$value.y
```

```{r}
df.cont.mhc = df %>%
  group_by(gene) %>%
  mutate(P = compute_contacts(chain, "CDR1", cdr1aa, "MHCI") + compute_contacts(chain, "CDR2", cdr2aa, "MHCI"))

df.cont.mhc$subset = "cd8"

df.cont.mhc.1 = df %>%
  group_by(gene) %>%
  mutate(P = compute_contacts(chain, "CDR1", cdr1aa, "MHCII") + compute_contacts(chain, "CDR2", cdr2aa, "MHCII"))

df.cont.mhc.1$subset = "cd4"

df.cont.mhc = rbind(df.cont.mhc, df.cont.mhc.1)
```

```{r}
df.cont.1 = df.cont.mhc %>%
  select(subset, gene, P)

colnames(df.cont.1) = c("subset", "variable", "contact.prob")

df.vusage = merge(df.vusage, df.cont.1)
```

```{r}
ggplot(df.vusage, aes(x=contact.prob, y=freq.ratio)) +
  geom_point(shape=21) +
  #geom_smooth() +
  geom_hline(yintercept = 1, color="red", linetype = "dashed") +
  facet_grid(subset~chain) +
  scale_y_log10() +
  theme_bw()
```

```{r}
a = aov(log10(freq.ratio) ~ subset + chain + contact.prob, subset(df.vusage, is.finite(freq.ratio) & freq.ratio > 0))
summary(a)
```

```{r}
ggplot(df.vusage, aes(x=contact.prob, y=value.x, color = subset)) +
  geom_point(shape=21) +
  facet_wrap(~chain, scales="free") 
```