---
title: "Hallmarks of TCR:peptide:MHC interactions inferred from structural data mining."
author: "Mikhail Shugay"
date: "February 29, 2016"
output: html_document
---

## Structural data used in the study

TCR:peptide:MHC complexe entries were obtained from PDB by a batch query with corresponding keywords. Complex records were then automatically annotated using in-house scripts that performed:

* TCR, MHC and antigen flags were assigned to chain records
* Antigen and host species were inferred
* MHC alleles were assigned using blast search against a database of MHC protein sequences manually assembled from public databases
* TCR partitioning into CDR and Framework regions was performed using custom IgBlast wrapper

This dataset was then used to generate a flat table with annotated TCR:antigen amino acid pairs.

```{r message = FALSE}
library(plyr)
library(ggplot2)
library(reshape2)
library(gplots)

df <- read.table("../result/structure.txt", header=T, sep="\t")
df$energy[is.na(df$energy)] <- 0

df <- ddply(df, .(tcr_v_allele), transform, 
            tcr_chain = factor(substr(as.character(tcr_v_allele[1]), 1, 3)))
```

The total number of complexes that were successfully annotated was

```{r comment=""}
length(levels(df$pdb_id))
```

and the total number of amino acid pairs was

```{r comment=""}
nrow(df)
```

## Selecting distance threshold for amino acid contacts

[GROMACS](http://www.gromacs.org/) software was used to calculate point energies for amino acid contacts using TCR:peptide:MHC structures. Each amino acid pair record in the database was then assigned with an interaction energy value using in-house scripts. Distances between residues were computed as the minimal distance between a pair of atoms using [Bio.PDB](http://biopython.org/DIST/docs/api/Bio.PDB-module.html) python package. Interaction energies grouped by CDR3 amino acid are plotted against residue distances below. Selected distance threshold for contacting residues is shown as a vertical line, the same value is used further throughout the manuscript.

```{r}
DIST_THRESHOLD = 6

ggplot(subset(df, distance <= 10 & energy <= 0), 
       aes(x=distance, y=-energy+1, colour=distance)) + 
  geom_point() + geom_vline(xintercept = DIST_THRESHOLD, linetype = "longdash") + 
  scale_y_log10(name = "Interaction energy (negative)") + xlab("Residue distance") +
  scale_colour_gradient(guide = FALSE, low = "#2c7fb8", high = "#f0f0f0") + 
  facet_wrap(~aa_tcr) + theme_bw()
```

## Studying distribution of TCR:antigen contact residues

### Number of TCR alpha and beta chain contacts is inversely correlated

The first hypothesis we've tested was the correlation between number of antigen contacts with TCR alpha and beta chains within the same TCR:peptide:MHC complex. We've focused on CDR3 region as the one having a critical role in antigen specificity and MHCI molecules. Based on previous observations of [Yokosuka et al. 2002](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2193687/) and [Turner et al. 1997](http://www.ncbi.nlm.nih.gov/pubmed/9278320), we have hypothesized that while the number of CDR3-antigen contacts per complex is quite stable, CDR3 regions of alpha and beta chains "compete" for antigen binding, resulting in TCRs in which one of the chains has a dominant role in antigen binding. Indeed, an inverse correlation between number of TCR alpha and beta CDR3-antigen contacts was observed as shown in the plot below.

```{r}
df.s <- ddply(subset(df, tcr_region == "CDR3" & mhc_type=="MHCI"), 
              .(pdb_id, tcr_chain),
              summarize,
              csum=sum(distance <= DIST_THRESHOLD), 
              len = mean(len_tcr),
              mhc_a_allele = mhc_a_allele[1]
              )
df.s <- subset(df.s, csum > 0)
df.s <- merge(subset(df.s, tcr_chain=="TRA"), subset(df.s, tcr_chain=="TRB"), 
              by="pdb_id", suffixes = c(".TRA",".TRB"))

ggplot(df.s, aes(x=csum.TRA / len.TRA, y=csum.TRB / len.TRB)) + 
  geom_smooth(method="lm", color="black") + 
  geom_point(size=4) + 
  geom_point(size=2, aes(color=csum.TRA-csum.TRB)) +
  scale_color_gradient2(guide = FALSE, low = "#a50026", mid = "#ffffbf", high = "#313695") + 
  scale_x_continuous(name = "TCR alpha chain CDR3-antigen contacts", limits=c(0,2)) + 
  scale_y_continuous(name = "TCR beta chain CDR3-antigen contacts", limits=c(0,2)) +
  #facet_wrap(~mhc_a_allele.TRA) +
  theme_bw()
```

The statistical significance of the observed dependency is given below.

```{r}
summary(lm(formula = csum.TRA ~ csum.TRB, data = df.s))
summary(lm(formula = csum.TRA / len.TRA ~ csum.TRB / len.TRB, data = df.s))
```

Distribution of total number of CDR3 contacts per complex is given below.

```{r}
df.s <- ddply(df.s, .(pdb_id),
              summarize, csum=csum.TRA+csum.TRB)
summary(df.s)
```

### TCR region and MHC contact preferences

Next, we've extended our analysis on other TCR regions (CDR1 and 2 in germline) and MHCII. Surprisingly, CDR1, but not CDR2, appears to confer a substantial number of TCR-antigen contacts, comparable to that of CDR3.

> Given the apparent role of CDR1 in peptide recognition, we present a hypothesis (yet to be tested) stating that the observed germline Variable (V) segment restriction and cross-reactivity to MHC alleles [Garcia 2012](http://www.ncbi.nlm.nih.gov/pubmed/22771140) can be at least partially explained by thymic selection due to the CDR1-encoded specificity to the self-peptide pool presented by a certain MHC. Note that this hypothesis is strengthened by the study of [Cole et al 2014](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3887192/) demonstrating that TCR-peptide specificity overrides affinity-enhancing TCR-MHC interactions.

Additionally, it appears that MHCI complexes have a higer number of TCR-antigen contacts than MHCII complexes. Our observations are summarized in the figure below.

```{r}
df.r <- ddply(df, .(pdb_id, tcr_chain, tcr_region, mhc_type), summarize, 
              csum=sum(distance <= DIST_THRESHOLD))

ggplot(subset(df.r, csum > 0), aes(x=tcr_region,group=tcr_region,y=csum,fill=tcr_region)) + 
  geom_boxplot() + xlab("TCR region") + ylab("Number of contacts") +
  scale_fill_brewer(guide = F, palette = "Set1") +
  facet_grid(tcr_chain~mhc_type) + theme_bw()
```

Statistical significance of results described in this section is provided below.

```{r}
a <- aov(csum~tcr_region + tcr_chain + mhc_type, df.r)
summary(a)
TukeyHSD(a, "mhc_type")
TukeyHSD(a, "tcr_region")
```

### Contact residues are tightly clustered on TCR and antigen length

We have next studied the distribution of interacting residues by their position on antigen and CDR sequences. Below are the plots of contact distribution grouped by TCR region conferring the contact (columns) and contact residue parent sequence (rows). CDR contacts tend to cluster near the center of corresponding region. Note the clear difference between TCR alpha and beta contacts that are closer to **N** and **C** terminus of the antigen peptide respectively.

```{r warning=FALSE}
df.p1 <- ddply(df, .(pdb_id, tcr_chain, tcr_region, mhc_type, pos_tcr), summarize,
              pos_norm = mean(pos_tcr - len_tcr / 2),
              contacts = sum(distance <= DIST_THRESHOLD))
df.p1$pos_tcr <- NULL
df.p1$sequence <- "TCR"

df.p2 <- ddply(df, .(pdb_id, tcr_chain, tcr_region, mhc_type, pos_antigen), summarize,
              pos_norm = mean(pos_antigen - len_antigen / 2),
              contacts = sum(distance <= DIST_THRESHOLD))
df.p2$pos_antigen <- NULL
df.p2$sequence <- "antigen"

df.p <- rbind(df.p1, df.p2)

ggplot(subset(df.p, contacts > 0), 
       aes(x=pos_norm, weight=contacts, fill=tcr_chain)) +
  geom_density(alpha=0.8) + 
  scale_fill_brewer(name = "TCR chain", palette = "Set1") + 
  ylab("Contact density") +
  scale_x_continuous(name = "Position, relative to region center", limits=c(-8,8)) +
  facet_grid(sequence~tcr_region) + 
  theme_bw()
```

Concat profile:

```{r}
library(stats)

df.pp <- subset(df.p, sequence == "TCR" & tcr_region == "CDR3")
wt <- ifelse(df.pp$contacts == 0, 0, 1)
wt <- wt / sum(wt)
d1 <- density(df.pp$pos_norm, weights = wt, from=-7, to=7, n=29)
plot(d1$x, d1$y)
print(data.frame(x=d1$x, y=d1$y))

wt <- df.pp$contacts
wt <- wt / sum(wt)
d2 <- density(df.pp$pos_norm, weights = wt, from=-7, to=7, n=29)
plot(d2$x, d2$y)
print(data.frame(x=d2$x, y=d2$y))
```

```{r}
df.pp <- ddply(subset(df.p, sequence == "TCR" & tcr_region == "CDR3"),
               .(pos_norm), summarize,
               total_contacts = sum(contacts))

df.pp$total_contacts <- df.pp$total_contacts / sum(df.pp$total_contacts)

print(df.pp)

ggplot(df.pp, aes(x=pos_norm, y=total_contacts)) + geom_bar(stat="identity")
```

> TODO: relative contact position density and CDR3 length

```{r}

df.p1 <- ddply(subset(df, tcr_region == "CDR3"), .(pdb_id, len_tcr, tcr_chain, mhc_type, pos_tcr), summarize,
              contacts = sum(distance <= DIST_THRESHOLD))

df.p1 <- ddply(df.p1, .(len_tcr, tcr_chain, mhc_type, pos_tcr), summarize,
               contacts = mean(contacts))

df.p1 <- ddply(df.p1, .(len_tcr, tcr_chain, mhc_type), transform,
               mean_contacts_pos = sum(pos_tcr * contacts / sum(contacts)))

ggplot(subset(df.p1, contacts > 0), 
       aes(y=len_tcr)) + 
  geom_point(aes(x=pos_tcr - len_tcr / 2, size = contacts, color=contacts)) + 
  geom_point(aes(x=mean_contacts_pos - len_tcr / 2), color="red", shape =3, size=3) +
  ylab("CDR3 length") +
  scale_x_continuous(name = "Position, relative to region center", limits=c(-7,7)) + 
  scale_color_gradient(name = "Mean #contacts", low="#7fcdbb", high="#2c7fb8") +
  scale_size(guide=F)+
  facet_grid(tcr_chain ~ mhc_type) + 
  theme_bw()
```

## Features of contacting amino acids

This section of the manuscript deals with amino acid features of TCR and antigen contact residues, such as physical properties and pairing preference. Positional information is not used, however, we restrict our analysis to CDR3 region.

### Physical properties of contacting TCR amino acids

The list of structural and basic physical properties of amino acids was taken from [Elhanati et al 2015](http://arxiv.org/pdf/1502.03136v2.pdf) (FIG. 13).

```{r}
aa_prop <- read.table("aa_properties.txt", header=T)

df.c <- subset(df, distance <= DIST_THRESHOLD & tcr_region == "CDR3")
df.c$tmp <- 1

# make sure all aa combinations here
tmp <- data.frame(expand.grid(aa_tcr = levels(aa_prop$aa), aa_antigen = levels(aa_prop$aa)))
tmp$count <- 0
df.c <- ddply(df.c, .(aa_tcr, aa_antigen), summarize, count = sum(tmp))
df.c <- rbind(df.c, tmp)
df.c <- ddply(df.c, .(aa_tcr, aa_antigen), summarize, count = sum(count))
```

We have observed a clear trend between number of contacts and certain properties of CDR3 amino acids. Notably, the **core** property derived in [Martin et al 2012](http://link.springer.com/article/10.1186%2F2046-1682-5-7) as the frequency of a given amino acid in the core region of protein-protein interacting interfaces had the best correlation with the number of contacts.

```{r, results = "hide", message=F, warning=F}
df.cc <- ddply(df.c, .(aa_tcr), summarize, count = sum(count))
colnames(df.cc) <- c("aa", "contacts")

df.cc2 <- merge(df.cc, subset(aa_prop, !(property %in% c("charge", "polar"))), all=T)

ggplot(df.cc2, aes(x=value,y=contacts)) + 
  geom_point() + geom_smooth() + ylab("Number of contacts") + xlab("Property value") +
  facet_wrap(~property, scales="free_x") + theme_bw()
```

Spearman correlation coefficients for each of these properties are given below.

```{r}
spearmentt <- function(r,n) r*sqrt((n-2)/(1-r^2))
calcpval <- function(r) {
  pval <- pt(spearmentt(r, 20), 20)
  min(pval, 1-pval)
}

df.cor <- ddply(df.cc2, .(property), summarize, r = cor(contacts, value, method="spearman"))
df.cor$pvalue <- sapply(df.cor$r, function(r) calcpval(r))

print(df.cor)
```
  
### Pairwise interaction preferences

Next, we've constructed and analyzed the CDR3-antigen amino acid contact matrix. The matrix is shown as a heatmap below, with row and column dendrograms constructed using hierarchical clustering (Ward's method). Note that here *rows* and *columns* corresponding to CDR3 and antigen amino acids respectively. 

The analysis results demonstrate clear clustering of CDR3 amino acids based on the **core** property, as shown by left color panel (*white* and *black* correspond to low and high **core** property values respectively). Notably, antigen amino acids were also nicely separated based on amino acid polarity, as shown top color panel (*black* corresponds to **polar** amino acids).

> **TODO** [Compare with MJ matrix](http://www.nature.com/articles/srep02919)

```{r}
df.cc <- dcast(df.c, aa_tcr ~ aa_antigen, fun.aggregate = mean, value.var="count")

contact_mat <- as.matrix(df.cc[,2:ncol(df.cc)])
rownames(contact_mat) <- df.cc$aa_tcr 

contact_mat[is.na(contact_mat)] <- 0

colgen <- function(aas, p, low, mid, hi) {
  panel <- colorpanel(100, low, mid, hi)
  df.1 <- merge(data.frame(aa=aas), subset(aa_prop, property == p))
  df.1$value <- as.integer((df.1$value - min(df.1$value)) / (max(df.1$value) - min(df.1$value)) * 99 + 1)
  return(panel[df.1$value])
}

heatmap.2(contact_mat, col=colorpanel(100, "#f0f0f0", "#7fcdbb", "#2c7fb8"),
          hclustfun = function(d) hclust(d, method="ward"),
          ColSideColors = colgen(colnames(contact_mat), "polar", "white", "white", "black"), 
          RowSideColors = colgen(rownames(contact_mat), "core", "white", "grey", "black"),
          density.info = "none", trace="none")
```

Raw number of contacts for CDR3 residues:

```{r}
rowSums(contact_mat)
```

Normalize contact matrix:

```{r}
contact_mat_norm <- contact_mat
rSum <- rowSums(contact_mat_norm)
cSum <- colSums(contact_mat_norm)
contact_mat_norm<-sweep(contact_mat_norm, 1, sqrt(rSum), `/`)
contact_mat_norm<-sweep(contact_mat_norm, 2, sqrt(cSum), `/`)
contact_mat_norm[is.na(contact_mat_norm)]<-0

heatmap.2(contact_mat_norm, col=colorpanel(100, "#f0f0f0", "#7fcdbb", "#2c7fb8"),
          hclustfun = function(d) hclust(d, method="ward"),
          ColSideColors = colgen(colnames(contact_mat), "polar", "white", "white", "black"), 
          RowSideColors = colgen(rownames(contact_mat), "core", "white", "grey", "black"),
          density.info = "none", trace="none")
```

Normalized number of contacts for CDR3 residues:

```{r}
rowSums(contact_mat_norm)
```

## Application of findings to HIV escape epitope data

Finally, we have applied our findings to study escape epitope variants from [NIAID HIV databases](http://www.hiv.lanl.gov/) database. The database was filtered to retain only variants with a single amino acid substitution. Escape variants were selected according to **E** (documented escape) and **NSF** (non-susceptible form) flags, and were further partitioned into three groups:

* **hla** variants that abrogate MHC binding (**DHB** flag, diminished HLA binding or increased off-rate)
* **tcr** variants that do not bind or show decreased binding to TCR (**TCR** flag).
* **other** unclassified escape variants

As only 8,9,10 and 11-mer epitopes had reported **tcr** escape variants, we've limited our analysis to this group.
                
```{r}
df.epi <- read.csv("ctl_variant.csv", header=T)

df.epi <- df.epi[grep("^[ACDEFGHIKLMNPQRSTVWY]\\d+[ACDEFGHIKLMNPQRSTVWY]$", df.epi$Mutation_epitope),]
df.epi <- df.epi[grep("^[ACDEFGHIKLMNPQRSTVWY]+$", df.epi$Epitope),]
len <- function(x) nchar(as.character(x))

df.epi$codes <- sapply(df.epi$Mutation_Type_Code, function(x) strsplit(as.character(x),", "))

isescape <- function(x) "E" %in% x || "LE" %in% x || "IE" %in% x || "NSF" %in% x

df.epi$type <- sapply(df.epi$codes, 
                      function(x) ifelse("TCR" %in% x, "tcr",
                                  ifelse("DHB" %in% x, "hla", 
                                  ifelse(isescape(x), "other", 
                                         "non-escape"
                                         ))))

df.epi <- subset(df.epi, type != "non-escape")

df.epi <- data.frame(seq = df.epi$Epitope,
                     aa_from = as.character(sapply(df.epi$Mutation_epitope, 
                                                   function(x) substr(x,1,1))),
                     aa_to = as.character(sapply(df.epi$Mutation_epitope, 
                                                 function(x) substr(x,len(x),len(x)))),
                     pos = sapply(df.epi$Mutation_epitope,
                                  function(x) as.integer(substr(x,2,len(x)-1))),
                     type = as.factor(df.epi$type),
                     elen = sapply(df.epi$Epitope,
                                   function(x) len(x))
                     )

df.epi <- unique(df.epi)
df.epi <- subset(df.epi, elen %in% c(8, 9, 10, 11))

df.epi$pos_adj <- apply(df.epi, 1, function(x) as.numeric(x[4])-(as.numeric(x[6]) / 2 + 1))

summary(df.epi)
```

As expected, **tcr**-mediate escape mutations are clustered in the central part of antigen (showing a two-peak picture resembling TCR contact positioning plot above), while **hla**-mediate escape mutations are clustered on **N** and **C** termini.

```{r}
ggplot(df.epi, aes(x=pos_adj, fill=type)) +
  geom_density(alpha=0.5) + 
  xlab("Position, relative to epitope center") +
  ylab("Variant density") +
  scale_fill_brewer(name = "Variant type", palette = "Set1") +
  scale_x_continuous(limits=c(-6,6)) + theme_bw()
```

Next, we have tested the hypothesis that **tcr**-mediated escape variants are biased in the direction of mutations that substantially change TCR recognition profile (*TRP*, i.e. a column from the contact amino acid matrix described above). We have therefore computed correlation coefficients between *TRPs* of original and substituted amino acids and compared them between **hla**, **tcr** and **other** mutation groups. We have also taken an advantage of positioning data and partitioned the **other** group into **other.tcr** and **other.hla** based on whether the variant was in *[-2, 2]* region in respect to antigen center or not.

```{r}
calc_dist <- function(row) cor(contact_mat[,row[2]], contact_mat[,row[3]])

df.epi$type2 <- as.factor(apply(df.epi, 1, function(x) ifelse(x[5] == "other", 
                                                  ifelse(abs(as.numeric(x[7])) <= 2, "other.tcr", "other.hla"),
                                                  x[5])))

summary(df.epi)

df.epi$dist <- apply(df.epi, 1, calc_dist)

ggplot(df.epi, aes(x = type2, fill=type2, group=type2, y=dist)) +
  geom_boxplot() + xlab("Variant type") + 
  ylab("Correlation of amino acid contact profiles") +
  scale_fill_brewer(guide=F, palette = "Set1") +
  theme_bw()
```

There was indeed a clear trend showing that **hla**-mediated escape mutations resulted in similar *TRPs*, while **tcr**-mediated mutations had least similar *TRPs*. Statistical details are given below.

```{r}
kruskal.test(dist ~ type2, df.epi)
wilcox.test(subset(df.epi, type2 == "other.hla")$dist, subset(df.epi, type2 == "other.tcr")$dist)
wilcox.test(subset(df.epi, type2 == "hla")$dist, subset(df.epi, type2 == "tcr")$dist)
```

The analysis described here can be further extended to immunogenic and non-immunogenic mutated self-peptides from cancer peptidomes once a database of cancer epitopes similar to [NIAID HIV databases](http://www.hiv.lanl.gov/) will become available.