---
title: "Hallmarks of TCR:peptide:MHC interactions inferred from structural data mining."
author: "Mikhail Shugay"
date: "August 15, 2016"
output: pdf_document
---

```{r}
# All imports

library(plyr)
library(ggplot2)
library(reshape2)
library(gplots)

# Helper functions

corrt <- function(r,n) r*sqrt((n-2)/(1-r^2))
calcpvalr <- function(r, n=20) {
  pval <- pt(corrt(r, n), n)
  min(pval, 1-pval)
}

p.annot <- function(p.val) {
  ifelse(p.val<0.001, "***", ifelse(p.val<0.01, "**", ifelse(p.val<0.05, "*", ifelse(p.val<0.1, "+", ""))))
}
```

## Structural data used in the study

TCR:peptide:MHC complexe entries were obtained from PDB by a batch query with corresponding keywords. Complex records were then automatically annotated using in-house scripts that performed:

* TCR, MHC and antigen flags were assigned to chain records
* Antigen and host species were inferred
* MHC alleles were assigned using blast search against a database of MHC protein sequences manually assembled from public databases
* TCR partitioning into CDR and Framework regions was performed using custom IgBlast wrapper

This dataset was then used to generate a flat table with annotated TCR:antigen amino acid pairs.

```{r message = FALSE}
df <- read.table("../result/structure.txt", header=T, sep="\t")
df$energy[is.na(df$energy)] <- 0

df <- ddply(df, .(tcr_v_allele), transform, 
            tcr_chain = factor(substr(as.character(tcr_v_allele[1]), 1, 3)))
```

The total number of complexes that were successfully annotated was

```{r comment=""}
length(levels(df$pdb_id))
```

and the total number of amino acid pairs was

```{r comment=""}
nrow(df)
```

## Selecting distance threshold for amino acid contacts

[GROMACS](http://www.gromacs.org/) software was used to calculate point energies for amino acid contacts using TCR:peptide:MHC structures. Each amino acid pair record in the database was then assigned with an interaction energy value using in-house scripts. Distances between residues were computed as the minimal distance between a pair of atoms using [Bio.PDB](http://biopython.org/DIST/docs/api/Bio.PDB-module.html) python package. Interaction energies grouped by CDR3 amino acid are plotted against residue distances below. Selected distance threshold for contacting residues is shown as a vertical line, the same value is used further throughout the manuscript.

```{r}
DIST_THRESHOLD = 6

ggplot(subset(df, distance <= 10 & energy <= 0), 
       aes(x=distance, y=-energy+1, colour=distance)) + 
  geom_point() + geom_vline(xintercept = DIST_THRESHOLD, linetype = "longdash") + 
  scale_y_log10(name = "Interaction energy (negative)") + xlab("Residue distance") +
  scale_colour_gradient(guide = FALSE, low = "#2c7fb8", high = "#f0f0f0") + 
  facet_wrap(~aa_tcr) + theme_bw()
```

## Studying distribution of TCR:antigen contact residues

### Number of TCR alpha and beta chain contacts is inversely correlated

The first hypothesis we've tested was the correlation between number of antigen contacts with TCR alpha and beta chains within the same TCR:peptide:MHC complex. We've focused on CDR3 region as the one having a critical role in antigen specificity and MHCI molecules. Based on previous observations of [Yokosuka et al. 2002](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2193687/) and [Turner et al. 1997](http://www.ncbi.nlm.nih.gov/pubmed/9278320), we have hypothesized that while the number of CDR3-antigen contacts per complex is quite stable, CDR3 regions of alpha and beta chains "compete" for antigen binding, resulting in TCRs in which one of the chains has a dominant role in antigen binding. Indeed, an inverse correlation between number of TCR alpha and beta CDR3-antigen contacts was observed as shown in the plot below.

```{r}
df.s <- ddply(subset(df, tcr_region == "CDR3"),
              .(pdb_id, tcr_chain),
              summarize,
              csum= sum(distance <= DIST_THRESHOLD), 
              len = mean(len_tcr)
              )

no_contact_pdb <- unique(with(subset(df.s, csum == 0), paste(pdb_id, tcr_chain)))

df.s <- subset(df.s, !(paste(pdb_id, tcr_chain) %in% no_contact_pdb))
df.s <- merge(subset(df.s, tcr_chain=="TRA"), subset(df.s, tcr_chain=="TRB"), 
              by="pdb_id", suffixes = c(".TRA",".TRB"))


ggplot(df.s, aes(x=csum.TRA, y=csum.TRB)) + 
  geom_smooth(method="lm", color="black") + 
  geom_point(size=4) + 
  geom_point(size=2, aes(color=csum.TRA-csum.TRB)) +
  scale_color_gradient2(guide = FALSE, low = "#a50026", mid = "#ffffbf", high = "#313695") + 
  scale_x_continuous(name = "TCR alpha contacts") + 
  scale_y_continuous(name = "TCR beta contacts") +  
  theme_bw()

# The statistical significance of the observed dependency is given below.

summary(lm(formula = csum.TRA ~ csum.TRB, data = df.s))
summary(lm(formula = csum.TRA / len.TRA ~ I(csum.TRB / len.TRB), data = df.s))
```

Distribution of total number of CDR3 contacts per complex is given below.

```{r}
summary(ddply(df.s, .(pdb_id), summarize, csum=csum.TRA+csum.TRB))
```

Number of antigen contacts is correlated with CDR3 length:

```{r}
df.s1 <- data.frame(csum = df.s$csum.TRA, len = df.s$len.TRA, chain="TRA")
df.s1 <- rbind(df.s1, data.frame(csum = df.s$csum.TRB, len = df.s$len.TRB, chain="TRB"))

ggplot(df.s1, aes(x=len, y=csum)) +
  geom_boxplot(aes(group=interaction(chain,len),fill=chain)) +
  #geom_smooth(method="lm", aes(color=chain)) +
  #geom_point(size=4, color="black") + 
  #geom_point(aes(color=chain), size=2, alpha=0.5) +
  scale_x_continuous("CDR3 length") + 
  scale_y_continuous("Number of CDR3 contacts") +
  scale_fill_brewer(palette = "Set1") + theme_bw()

summary(lm(formula = csum.TRA ~ len.TRA, data = df.s))
summary(lm(formula = csum.TRB ~ len.TRB, data = df.s))
```

### TCR region and MHC contact preferences

Next, we've extended our analysis on other TCR regions (CDR1 and 2 in germline) and MHCII. Surprisingly, CDR1, but not CDR2, appears to confer a substantial number of TCR-antigen contacts, comparable to that of CDR3.

> Given the apparent role of CDR1 in peptide recognition, we present a hypothesis (yet to be tested) stating that the observed germline Variable (V) segment restriction and cross-reactivity to MHC alleles [Garcia 2012](http://www.ncbi.nlm.nih.gov/pubmed/22771140) can be at least partially explained by thymic selection due to the CDR1-encoded specificity to the self-peptide pool presented by a certain MHC. Note that this hypothesis is strengthened by the study of [Cole et al 2014](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3887192/) demonstrating that TCR-peptide specificity overrides affinity-enhancing TCR-MHC interactions.

Additionally, it appears that MHCI complexes have a higer number of TCR-antigen contacts than MHCII complexes. Our observations are summarized in the figure below.

```{r}
df.r <- ddply(df, .(pdb_id, tcr_chain, tcr_region, mhc_type), summarize, 
              csum=sum(distance <= DIST_THRESHOLD))

ggplot(subset(df.r, csum > 0), aes(x=tcr_region,group=tcr_region,y=csum)) + 
  geom_violin() +
  geom_boxplot(aes(fill=tcr_region),width=0.1) +
  xlab("TCR region") + ylab("Number of contacts") +
  scale_fill_brewer(guide = F, palette = "Set1") +
  facet_grid(tcr_chain~mhc_type) + theme_bw()
```

Statistical significance of results described in this section is provided below.

```{r}
a <- aov(csum~tcr_region + tcr_chain + mhc_type, df.r)
summary(a)
TukeyHSD(a, "mhc_type")
TukeyHSD(a, "tcr_region")

#check for CDR3 only
a <- aov(csum~tcr_chain + mhc_type, subset(df.r, tcr_region == "CDR3"))
TukeyHSD(a, "mhc_type")
```

### Contact residues are tightly clustered on TCR and antigen length

We have next studied the distribution of interacting residues by their position on antigen and CDR sequences. Below are the plots of contact distribution grouped by TCR region conferring the contact (columns) and contact residue parent sequence (rows). CDR contacts tend to cluster near the center of corresponding region. Note the clear difference between TCR alpha and beta contacts that are closer to **N** and **C** terminus of the antigen peptide respectively.

```{r warning=FALSE}
df.p1 <- ddply(df, .(pdb_id, tcr_chain, tcr_region, mhc_type, pos_tcr), summarize,
              pos_norm = mean(pos_tcr - len_tcr / 2),
              contacts = sum(distance <= DIST_THRESHOLD))
df.p1$pos_tcr <- NULL
df.p1$sequence <- "TCR"

df.p2 <- ddply(df, .(pdb_id, tcr_chain, tcr_region, mhc_type, pos_antigen), summarize,
              pos_norm = mean(pos_antigen - len_antigen / 2),
              contacts = sum(distance <= DIST_THRESHOLD))
df.p2$pos_antigen <- NULL
df.p2$sequence <- "antigen"

df.p <- rbind(df.p1, df.p2)

ggplot(subset(df.p, contacts > 0), 
       aes(x=pos_norm, weight=contacts, fill=tcr_chain)) +
  geom_density(alpha=0.8) + 
  scale_fill_brewer(name = "TCR chain", palette = "Set1") + 
  ylab("Contact density") +
  scale_x_continuous(name = "Position, relative to region center", limits=c(-8, 8)) +
  facet_grid(sequence~tcr_region) + 
  theme_bw()
```

Relative contact position density and CDR3 length.

```{r}
df.p1 <- ddply(subset(df, tcr_region == "CDR3"), .(pdb_id, len_tcr, tcr_chain, mhc_type, pos_tcr), summarize,
              contacts = sum(distance <= DIST_THRESHOLD))

df.p1 <- ddply(df.p1, .(len_tcr, tcr_chain, mhc_type, pos_tcr), summarize,
               contacts = mean(contacts))

df.p1 <- ddply(df.p1, .(len_tcr, tcr_chain, mhc_type), transform,
               mean_contacts_pos = sum(pos_tcr * contacts / sum(contacts)))

ggplot(subset(df.p1, contacts > 0), 
       aes(y=len_tcr)) + 
  geom_point(aes(x=pos_tcr - len_tcr / 2, size = contacts, color=contacts)) + 
  geom_point(aes(x=mean_contacts_pos - len_tcr / 2), color="red", shape =3, size=3) +
  ylab("CDR3 length") +
  scale_x_continuous(name = "Position, relative to region center", limits=c(-7,7)) + 
  scale_color_gradient(name = "Mean #contacts", low="#7fcdbb", high="#2c7fb8") +
  scale_size(guide=F)+
  facet_grid(tcr_chain ~ mhc_type) + 
  theme_bw()
```

Contour map for contacts between alpha-beta CDR3 and antigen. Dashed lines show mean contact position in CDR3 (vertical) and antigen (horizontal), shaded areas show +/- SD.

```{r}
df.p2 <- ddply(subset(df, tcr_region == "CDR3" & distance <= DIST_THRESHOLD), 
               .(pdb_id, tcr_chain, pos_tcr, pos_antigen), summarize,
               pos_norm_tcr = 2 * (pos_tcr - len_tcr / 2) / len_tcr,
               pos_norm_antigen = 2 * (pos_antigen - len_antigen / 2) / len_antigen,
               odd = len_tcr %% 2 == 0)

df.p3 <- ddply(df.p2, .(tcr_chain), summarize,
               pos_mean_tcr = mean(pos_norm_tcr),
               pos_sd_tcr = sd(pos_norm_tcr),
               pos_mean_antigen = mean(pos_norm_antigen),
               pos_sd_antigen = sd(pos_norm_antigen))

ggplot(df.p2, aes(x=pos_norm_tcr, y=pos_norm_antigen)) +
  stat_density_2d(aes(fill = ..level..), geom = "polygon", alpha=0.3) +
  geom_density_2d(aes(color=tcr_chain)) + 
  scale_color_brewer(palette = "Set1") +
  annotate(geom="rect", ymin=df.p3[1,4]-df.p3[1,5], ymax=df.p3[1,4]+df.p3[1,5], xmin=-Inf, xmax=+Inf, 
           fill="red", alpha=0.1) +
  geom_hline(yintercept=df.p3[1,4], color="red", linetype="dashed") +
  annotate(geom="rect", ymin=df.p3[2,4]-df.p3[2,5], ymax=df.p3[2,4]+df.p3[2,5], xmin=-Inf, xmax=+Inf, 
           fill="blue", alpha=0.1) +
  geom_hline(yintercept=df.p3[2,4], color="blue", linetype="dashed") +
  annotate(geom="rect", xmin=df.p3[1,2]-df.p3[1,3], xmax=df.p3[1,2]+df.p3[1,3], ymin=-Inf, ymax=+Inf, 
           fill="red", alpha=0.1) +
  geom_vline(xintercept=df.p3[1,2], color="red", linetype="dashed") +
  annotate(geom="rect", xmin=df.p3[2,2]-df.p3[2,3], xmax=df.p3[2,2]+df.p3[2,3], ymin=-Inf, ymax=+Inf, 
           fill="blue", alpha=0.1) +
  geom_vline(xintercept=df.p3[2,2], color="blue", linetype="dashed") +
  annotate(geom="point", x=df.p3[1,2], y=df.p3[1,4],
           color="red") +
  annotate(geom="point", x=df.p3[2,2], y=df.p3[2,4],
           color="blue") +
  #annotate(geom="rect", ymin=df.p3[1,4]-df.p3[1,5], ymax=df.p3[1,4]+df.p3[1,5], xmin=-Inf, xmax=+Inf, 
  #         fill="red", alpha=0.5) +
  scale_fill_gradient(low="white", high="black") + 
  scale_x_continuous("Relative position in CDR3", limits=c(-1,1))+
  scale_y_continuous("Relative position in antigen", limits=c(-1,1))+
  theme_bw()
```

Contour plots of CDR3:antigen contact positions for each complex, colored by chain. Linear fitting is performed to bulk contact position plot showing some degree of dependency between CDR3 and antigen position.

```{r}
library(scales)

ggplot(df.p2, aes(x=pos_norm_tcr, y=pos_norm_antigen, color=tcr_chain)) + 
  geom_density_2d(size=0.1) +
  geom_point(size=0.1) +
  facet_wrap(~pdb_id) + 
  scale_x_continuous("Relative position in CDR3", limits=c(-1,1), oob = rescale_none)+
  scale_y_continuous("Relative position in antigen", limits=c(-1,1), oob = rescale_none) +
  scale_color_brewer(palette = "Set1") +
  scale_fill_brewer(palette = "Set1") +
  theme_bw() +
  theme(axis.line=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank())

summary(lm(pos_norm_antigen ~ pos_norm_tcr * tcr_chain, df.p2))

ggplot(df.p2, aes(x=pos_norm_tcr, y=pos_norm_antigen, color=tcr_chain)) + 
  geom_point(shape = 21) + geom_smooth(method="lm") +
  scale_x_continuous("Relative position in CDR3", limits=c(-1,1), oob = rescale_none)+
  scale_y_continuous("Relative position in antigen", limits=c(-1,1), oob = rescale_none) +
  scale_color_brewer(palette = "Set1") +
  theme_bw()
```

## Features of CDR3 residues that explain number of antigen contacts they form

This section of the manuscript deals with amino acid features of TCR and antigen contact residues, such as physical properties and pairing preference. Positional information is not used, however, we restrict our analysis to CDR3 region.

```{r}
df.0 <- subset(df, tcr_region == "CDR3" & !(paste(pdb_id, tcr_chain) %in% no_contact_pdb)) # also remove chains with 0 contacts
```

Most variance in the number of contacts of a given amino acid is explained by its positioning:

```{r, results = "hide", message=F, warning=F}
pos_mean_tcr <- df.p3$pos_mean_tcr
names(pos_mean_tcr) <- df.p3$tcr_chain

calc_pos_norm <- function(pos_tcr, len_tcr, tcr_chain) {
  half_len <- len_tcr[1] / 2
  ((pos_tcr - half_len) / half_len) ^ 2
}

df.c <- ddply(df.0, .(pdb_id, tcr_chain, aa_tcr, pos_tcr), summarize,
              pos_norm = calc_pos_norm(pos_tcr, len_tcr)[1],
              contacts = sum(distance <= DIST_THRESHOLD),
              count = 1)

aovsum <- summary(aov(contacts ~ aa_tcr * pos_norm, subset(df.c, aa_tcr != "C")))

var <- aovsum[[1]]["Sum Sq"]

print("Percent of variance explained")

print(var / sum(var) * 100)

aovres <- aov(pos_norm ~ aa_tcr, subset(df.c, aa_tcr != "C"))
summary(aovres)
```

Illustrate the above by plotting mean number of contacts vs CDR3 AA frequency at each relative CDR3 position.

```{r}
df.c1 <- ddply(subset(df.c, aa_tcr != "C"), .(aa_tcr, pos_norm), summarize,
              contacts_sum = sum(contacts),
              count = sum(count))

ggplot(df.c1, aes(x=pos_norm)) +  
  geom_density(aes(weight = contacts_sum), fill="red", color=NA) +
  geom_density(aes(weight = count), fill=NA, linetype="dashed") +
  #scale_x_continuous("Distance from CDR3 center", limits=c(0,1)) +
  ylab("") +
  facet_wrap(~aa_tcr, scales="free_y")+
  theme_bw() +
  theme(axis.line=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.text.x = element_text(angle = 60, hjust = 1))
```

### CDR3 AA features and number of contacts

The list of structural and basic physical properties of amino acids was taken from [Elhanati et al 2015](http://arxiv.org/pdf/1502.03136v2.pdf) (FIG. 13). 

```{r}
aa_prop <- read.table("aa_properties.txt", header=T)
```

None of aforementioned AA properties is correlated with mean number of contacts.

```{r}
df.cc <- ddply(df.c, .(aa_tcr), summarize,
               contacts = sum(contacts),
               frequency = sum(count))
df.cc <- merge(df.cc, aa_prop, by.x = "aa_tcr", by.y="aa")

# Mean number of contacts vs property value

ggplot(df.cc, aes(x=value, y=contacts / frequency)) + 
  geom_point() + geom_smooth(color="#377eb8", method="lm") +
  geom_text(aes(label=aa_tcr), color="#e41a1c", vjust=-0.5) +
  ylab("Mean number of contacts") + xlab("Property value") +
  facet_wrap(~property, scales="free_x") + theme_bw() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

# correlation with mean number of contacts
df.cor <- ddply(subset(df.cc, !(property %in% c("polar"))), 
                       .(property), summarize, r = cor(contacts / frequency, value, method="pearson"))
df.cor$pvalue <- sapply(df.cor$r, function(r) calcpvalr(r))
df.cor$pvalue <- p.adjust(df.cor$pvalue)
df.cor$p.annot <- p.annot(df.cor$pvalue)

print(df.cor)

# check polar separately
t.test(contacts / frequency ~ value, data = subset(df.cc, property == "polar"))
```

Overall, these results suggest that AA placement in CDR3 is definitive for the number of contacts and this effect should be corrected for to account for individual AA features.

### CDR3 AA contact number factors inferred using positional model

Fit total number of contacts of a given CDR3 AA using GLM. The formula is 

``contacts ~ a(aa_tcr) * exp(-b(aa_tcr) * (pos_tcr - len_tcr/2)^2 / (len_tcr/2)^2)``

```{r}
# Compute the table of number of CDR3 contacts
df.contpos1 <- ddply(df.0, .(pdb_id, tcr_chain, pos_tcr, aa_tcr, len_tcr), summarize,
              pos_norm = calc_pos_norm(pos_tcr, len_tcr, tcr_chain)[1],
              contacts = sum(distance <= DIST_THRESHOLD))

# Strip Cys at all
df.contpos1 <- subset(df.contpos1, aa_tcr != "C")
```

Run the fitting with GLM and Poisson distribution family (somewhat time consuming):

```{r}
fit.c1 <- glm(contacts ~ aa_tcr + aa_tcr : pos_norm - 1, df.contpos1, 
           family=poisson())

summary(fit.c1)
```

Check the fitting for the number of contacts for each of individual cases:

```{r}
df.contpos1$contacts.fit <- fitted.values(fit.c1)

summary(aov(contacts ~ contacts.fit, df.contpos1))
summary(lm(contacts ~ contacts.fit, df.contpos1))

ggplot() +
  geom_violin(data=df.contpos1, 
              aes(x=contacts, y=contacts.fit, group = contacts), fill = "#74a9cf", width=1) + 
  geom_point(data=ddply(df.contpos1, .(contacts), summarize, med=median(contacts.fit)), 
             aes(x=contacts, y=med), color="red", size=15, shape=95) +
  xlab("Number of antigen contacts") + ylab("Estimated value") +
  theme_bw()
```

Check the fitting for total sum of contacts in each CDR3:

```{r}
df.contpos1.sum <- ddply(df.contpos1, .(pdb_id, tcr_chain, len_tcr),
                        summarize,
                        contacts.sum = sum(contacts),
                        contacts.fit.sum = sum(contacts.fit))

ggplot(df.contpos1.sum, aes(contacts.sum, contacts.fit.sum)) +   
  geom_errorbar(aes(ymin = contacts.fit.sum - sqrt(contacts.fit.sum / len_tcr),
                    ymax = contacts.fit.sum + sqrt(contacts.fit.sum / len_tcr))) +
  geom_smooth(method="lm", color="black") + 
  geom_point(shape=21, size = 4, aes(fill = tcr_chain)) +
  scale_y_continuous("Estimated value") +
  scale_x_continuous("Total number of CDR3 contacts") +
  scale_fill_brewer("TCR chain", palette = "Set1") +
  theme_bw()

summary(lm(contacts.fit.sum ~ contacts.sum, df.contpos1.sum))
summary(aov(contacts.fit.sum ~ contacts.sum, df.contpos1.sum))
```

Now lets analyze the resulting coefficients for each amino acids. First, prepare coefficient table.

```{r}
df.coef1 <- as.data.frame(summary(fit.c1)$coefficients)

df.coef1$type <- ifelse(grepl("pos_norm", rownames(df.coef1)), "b", 
                        ifelse(grepl("aa_tcr", rownames(df.coef1)), "a", "c"))
df.coef1$aa <- sapply(rownames(df.coef1), function(x) strsplit(x,"")[[1]][7])

rownames(df.coef1) <- NULL
colnames(df.coef1) <- c("value", "std", "z", "P", "type", "aa")
df.coef1$z <- NULL
df.coef1$P <- NULL
```

Compute Z scores and P-values with respect to baseline (average coefficient value).

```{r}
df.coef1.plt <- subset(df.coef1, type == "a")
df.coef1.plt$value <- df.coef1.plt$value - mean(df.coef1.plt$value)

# transform mean and std back
df.coef1.plt$value <- exp(df.coef1.plt$value)
df.coef1.plt$std <- df.coef1.plt$value * df.coef1.plt$std

# compute P-values
df.coef1.plt$z <- with(df.coef1.plt, (value - 1) / std)
df.coef1.plt$p <- 2*pnorm(-abs(df.coef1.plt$z))
df.coef1.plt$p.adj <- p.adjust(df.coef1.plt$p)
df.coef1.plt$p.txt <- p.annot(df.coef1.plt$p.adj)
df.coef1.plt
```

Plot base coefficients (``a(tcr_aa)``):

```{r}
df.coef1.plt$strong <- as.factor(ifelse(df.coef1.plt$aa %in% c("C","F","I","L","M","V","W","Y"), "yes", "no"))
df.coef1.plt$strong <- relevel(df.coef1.plt$strong, "yes")

# order AAs based on value
df.coef1.plt$aa = with(df.coef1.plt, factor(aa, levels=aa[order(-value)], ordered=TRUE))

ggplot(df.coef1.plt) +   
  geom_bar(aes(x=aa, y=value, fill=strong), stat="identity", color="black", width=0.7) + 
  geom_hline(yintercept = 1, color="black", linetype="longdash") + 
  #geom_point(aes(x=aa, y=-0.1, color=strong), size=8, shape=15) +
  geom_errorbar(aes(x=aa, ymin=value, ymax=value+std), width=0.4) +
  geom_text(aes(x=aa, y = value+std + 0.05, label=p.txt)) +
  xlab("") + scale_y_continuous("",limits=c(-0, 3.5), expand = c(0,0)) +
  scale_fill_brewer("Strongly interacting (MJ)", palette = "Set1") +
  theme_bw()
```

Correlation of coefficients with AA features:

```{r}
coef.aa_prop <- data.frame(aa=unique(aa_prop$aa))
coef.aa_prop$coeff <- sapply(coef.aa_prop$aa, function(x) coefficients(fit.c1)[paste("aa_tcr",x,sep = "")])
coef.aa_prop <- merge(coef.aa_prop, aa_prop)
coef.aa_prop <- subset(coef.aa_prop, !(aa %in% c("C")))

ggplot(coef.aa_prop, aes(x=value, y=exp(coeff))) + 
  geom_point() + geom_smooth(color="#377eb8", method="lm") +
  geom_text(aes(label=aa), color="#e41a1c", vjust=-0.5) +
  ylab("AA coefficient") + xlab("Property value") +
  facet_wrap(~property, scales="free_x") + theme_bw() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

df.cor <- ddply(subset(coef.aa_prop, !(property %in% c("charge", "polar"))), 
                       .(property), summarize, r = cor(exp(coeff), value, method="pearson"))

df.cor$pvalue <- sapply(df.cor$r, function(r) calcpvalr(r, 19))
df.cor$pvalue <- p.adjust(df.cor$pvalue)

print(df.cor)
```

### Analyze TCRs with known antigen specificity using the model

The following function is used to compute the number of contacts using the trained model:

```{r}
mdl.coef.df <- as.data.frame(summary(fit.c1)$coefficients)
mdl.coef <- mdl.coef.df$Estimate
names(mdl.coef) <- rownames(mdl.coef.df)

calc_contact_num <- function(aa_tcr, pos_tcr, len_tcr, tcr_chain) {
  pos_norm <- calc_pos_norm(pos_tcr, len_tcr)
  exp(mapply(function(x, y, z) ifelse(x == "C", -Inf, 
                                           mdl.coef[paste("aa_tcr", x, sep = "")] + 
                                           mdl.coef[paste("aa_tcr", x, ":pos_norm", sep = "")] * y
           ), aa_tcr, pos_norm, len_tcr))
}

calc_contact_num_sum <- function(seq) {
  aa_tcr <- strsplit(seq, "")[[1]]
  len_tcr <- length(aa_tcr)
  ifelse(len_tcr == 0, 0,
         sum(calc_contact_num(aa_tcr, 1:len_tcr, len_tcr)))
}

#calc_contact_num_sum("CASSLAPGATNEKLFF")
```

Load paired TCR alpha beta data for *RPRGEVRFL* epitope from **PMID:20139278**:

```{r}
df.hsv2abpairs <- read.table("datasets/HSV-2_tcr_ab_pairs.txt", header = T, sep="\t", stringsAsFactors = F)

df.hsv2abpairs <- ddply(df.hsv2abpairs, .(cdr3.alpha, cdr3.beta), transform,
                  len.alpha = nchar(cdr3.alpha),
                  len.beta = nchar(cdr3.beta),
                  contacts.alpha = calc_contact_num_sum(cdr3.alpha),
                  contacts.beta = calc_contact_num_sum(cdr3.beta))

ggplot(df.hsv2abpairs, aes(x=contacts.alpha, y=contacts.beta)) + 
  geom_smooth(method="lm", color="black") + 
  geom_errorbar(aes(ymin = contacts.beta - sqrt(contacts.beta / len.beta), 
                    ymax = contacts.beta + sqrt(contacts.beta / len.beta))) +
  geom_errorbarh(aes(xmin = contacts.alpha - sqrt(contacts.alpha / len.alpha), 
                    xmax = contacts.alpha + sqrt(contacts.alpha / len.alpha))) +
  geom_point(size=4, aes(fill=contacts.alpha-contacts.beta), shape=21) +
  scale_fill_gradient2(guide = FALSE, low = "#a50026", mid = "#ffffbf", high = "#313695") + 
  scale_x_continuous(name = "TCR alpha contacts est.") + 
  scale_y_continuous(name = "TCR beta contacts est.") +
  theme_bw()

summary(lm(contacts.beta ~ contacts.alpha, df.hsv2abpairs))
# normalized by length
summary(lm(contacts.beta / len.beta ~ I(contacts.alpha / len.alpha), df.hsv2abpairs))
```

SIV-specific TCRs from David Price et al. monkey study. Differences between CM9- and TL8-specific TCRs:

```{r}
df.tcr_siv <- read.table("datasets/siv_tcrs.txt", header=T, stringsAsFactors = F)

df.tcr_siv <- ddply(df.tcr_siv, .(cdr3), transform,
                  len = nchar(cdr3),
                  contacts = calc_contact_num_sum(cdr3))

df.tcr_siv <- subset(df.tcr_siv, len >= 12 & len <= 18)

ggplot(df.tcr_siv,
       aes(x=len, y = contacts, color=antigen)) +
  geom_smooth(method="lm") +
  geom_boxplot(aes(group=interaction(len, antigen)), width=0.4) + 
  scale_color_brewer(palette = "Set1") + 
  scale_x_continuous("CDR3 length", breaks=12:18) + 
  ylab("Estimated number of contacts") +
  theme_bw()

summary(aov(contacts ~ len + antigen, df.tcr_siv))

df.pvals <- data.frame(len=12:18)
df.pvals$pvals <- sapply(df.pvals$len, function(x) wilcox.test(contacts ~ antigen,
                                                               subset(df.tcr_siv, len == x), exact=F)[[3]])
print(df.pvals)
```

Comparing contact profiles for each CDR3 length between two epitopes:

```{r}
df.tcr_siv.flat <- data.frame(cdr3.beta = character(), aa_tcr = character(),
                           pos_tcr = integer(), 
                           antigen.species = character(), antigen.epitope = character())

for (i in 1:nrow(df.tcr_siv)) {
  .row <- df.tcr_siv[i, ]
  .aa_tcr <- strsplit(.row$cdr3, split="")[[1]]
  .df.ext <- data.frame(cdr3 = .row$cdr3, 
                        aa_tcr = .aa_tcr, 
                        pos_tcr = 1:length(.aa_tcr),
                        len_tcr = length(.aa_tcr),
                        antigen = .row$antigen)
  df.tcr_siv.flat <- rbind(df.tcr_siv.flat, .df.ext)
}

df.tcr_siv.flat$contacts <- with(df.tcr_siv.flat, 
                                 mapply(function(x,y,z) calc_contact_num(x,y,z), aa_tcr, pos_tcr, len_tcr))

get_contacts <- function(pos, len) {
  subset(df.tcr_siv.flat, len_tcr == len[1] & pos_tcr == pos[1] & antigen == "CTPYDINQM")$contacts
}

df.tcr_siv.plt <- ddply(subset(df.tcr_siv.flat, len_tcr %in% c(14:18)),
                        .(pos_tcr, len_tcr, antigen), summarize,
                        contacts.mean=median(contacts),
                        pval=wilcox.test(contacts, get_contacts(pos_tcr, len_tcr))[[3]])

df.tcr_siv.plt$ptxt <- p.annot(ifelse(is.nan(df.tcr_siv.plt$pval), 1, df.tcr_siv.plt$pval))

ggplot(df.tcr_siv.plt, aes(x=pos_tcr, y=contacts.mean,
                         color=antigen)) + 
  geom_line() +
  geom_point() +
  geom_text(aes(y = contacts.mean+0.2, label = ptxt)) +
  scale_color_brewer(palette = "Set1") + 
  ylab("Contacts estimate") +
  scale_x_continuous("", breaks=1:18) +
  facet_grid(len_tcr~., scales="free_x") +
  theme_bw()
```

### Analysis of T-cell subsets


CD4+CD25- and CD4+CD25+ subset with effector T-reg phenotype from prof. Gaston:

```{r}
read_annot_table <- function(.path, .donor,.subset,.sample) {
  fname <- paste(.path, paste(.donor,.subset,.sample,"txt",sep="."), sep="")
  if(file.exists(fname)) {
    .df <- read.table(fname, header=T, sep="\t", fill=T, stringsAsFactors = F)
    return (with(.df, data.frame(cdr3contact= aaprop.cdr3contact, weight=freq,
             donor = .donor, subset = .subset, sample = .sample)))
  }
}

df.treg <- data.frame(cdr3contact = numeric(), donor=factor(), subset = factor(), sample = integer())

for (donor in c("steve", "tony", "kevin")) {
  for (s in c("G2", "Tresp")) {
    for (replica in 1:3) {
      df.treg <- rbind(df.treg, read_annot_table("datasets/treg/", donor, s, replica))
    }
  }
}

ggplot(df.treg, aes(x=cdr3contact, color=subset, linetype=donor)) +
  stat_ecdf(geom="line") + 
  scale_color_manual(values = c("red", "black")) +
  scale_x_continuous(limits=c(10, 20), expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0))+
  theme_bw()

ks.test(subset(df.treg, donor == "steve" & subset=="Tresp")$cdr3contact,
        subset(df.treg, donor == "steve" & subset=="G2")$cdr3contact)
ks.test(subset(df.treg, donor == "tony" & subset=="Tresp")$cdr3contact,
        subset(df.treg, donor == "tony" & subset=="G2")$cdr3contact)
ks.test(subset(df.treg, donor == "kevin" & subset=="Tresp")$cdr3contact,
        subset(df.treg, donor == "kevin" & subset=="G2")$cdr3contact)
```

Memory T-cells from from prof. Goronzy aging study:

```{r}
df.mem <- data.frame(cdr3contact = numeric(), donor=factor(), subset = factor(), sample = integer())

for (donor in c("Donor7", "Donor9")) {
  for (s in c("M", "CM", "EM", "TEM")) {
    for (replica in 1:5) {
      df.mem <- rbind(df.mem, read_annot_table("datasets/memory/", donor, s, replica))
    }
  }
}

ggplot(df.mem, aes(x=cdr3contact, color=subset, linetype=subset)) +
  stat_ecdf(geom="line") + 
  scale_color_manual(values=c("black", "blue", "orange", "red")) +
  scale_linetype_manual(values = c("dashed", "dashed", "dashed", "solid")) +
  scale_x_continuous(limits=c(10, 20), expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0))+
  facet_grid(donor~.) +
  theme_bw()

for (d in c("Donor7", "Donor9")) {
  for (s in c("M", "CM", "EM")) {
    print(paste(d,s))
    print(ks.test(subset(df.mem, donor == d & subset==s)$cdr3contact,
            subset(df.mem, donor == d & subset=="TEM")$cdr3contact))
  }
}
```

## Application of findings to HIV escape epitope data

### Pairwise contact preferences

```{r}
df.cm <- ddply(df.0, .(aa_tcr, aa_antigen), summarize,
               contacts_sum = sum(distance <= DIST_THRESHOLD),
               total = length(distance))

contact_mat <- dcast(df.cm, aa_tcr ~ aa_antigen, value.var = "contacts_sum")
contact_mat[is.na(contact_mat)] <- 0
rownames(contact_mat) <- contact_mat[,1]
contact_mat[,1] <- NULL
contact_mat <- contact_mat[-2, ]
contact_mat <- as.matrix(contact_mat)

colgen <- function(aas, p, low, mid, hi) {
  panel <- colorpanel(100, low, mid, hi)
  df.1 <- merge(data.frame(aa=aas), subset(aa_prop, property == p))
  df.1$value <- as.integer((df.1$value - min(df.1$value)) / (max(df.1$value) - min(df.1$value)) * 99 + 1)
  return(panel[df.1$value])
}

cn <- colnames(contact_mat)
colgen_cols <- ifelse(cn %in% c("A", "C", "I", "L", "M", "F", "W", "V"), "blue",
                  ifelse(cn %in% c("G", "H", "P", "S", "T", "Y"), "red", "yellow"))    

heatmap.2(contact_mat,
          col=colorpanel(100, "white", "#74a9cf", "#045a8d"),
          #distfun = function(x) as.dist(1-cor(t(x))/2),
          #hclustfun = function(d) hclust(d, method="ward"),
          #ColSideColors = colgen_cols,
          #RowSideColors = colgen(rownames(mat.cm), "core", "white", "grey", "black"),
          density.info = "none", trace="none")
```

```{r}
calc_dist_1 <- function(x, y) cor(contact_mat[,x], contact_mat[,y])

df.ag.sim <- expand.grid(colnames(contact_mat), colnames(contact_mat))

df.ag.sim$r <- with(df.ag.sim, mapply(function(x,y) calc_dist_1(x, y), Var1, Var2))

mat.ag.sim <- dcast(df.ag.sim, Var1~Var2)
rownames(mat.ag.sim) <- mat.ag.sim[,1]
mat.ag.sim[,1] <- NULL
mat.ag.sim <- as.matrix(mat.ag.sim)

heatmap.2(mat.ag.sim,
          col=colorpanel(100, "#ffffb2", "#fd8d3c", "#bd0026"),
         # distfun = function(x) as.dist(1-cor(t(x))/2),
         # hclustfun = function(d) hclust(d, method="ward"),
         # ColSideColors = colgen_cols,
          #RowSideColors = colgen(rownames(mat.cm), "hydrop", "white", "grey", "black"),
          density.info = "none", trace="none")
```

### Analyzis of NIAID HIV db

Finally, we have applied our findings to study escape epitope variants from [NIAID HIV databases](http://www.hiv.lanl.gov/) database. The database was filtered to retain only variants with a single amino acid substitution. Escape variants were selected according to **E** (documented escape) and **NSF** (non-susceptible form) flags, and were further partitioned into three groups:

* **hla** variants that abrogate MHC binding (**DHB** flag, diminished HLA binding or increased off-rate)
* **tcr** variants that do not bind or show decreased binding to TCR (**TCR** flag).
* **other** unclassified escape variants

As only 8,9,10 and 11-mer epitopes had reported **tcr** escape variants, we've limited our analysis to this group.
                
```{r}
df.epi <- read.csv("datasets/ctl_variant.csv", header=T)

df.epi <- df.epi[grep("^[ACDEFGHIKLMNPQRSTVWY]\\d+[ACDEFGHIKLMNPQRSTVWY]$", df.epi$Mutation_epitope),]
df.epi <- df.epi[grep("^[ACDEFGHIKLMNPQRSTVWY]+$", df.epi$Epitope),]
len <- function(x) nchar(as.character(x))

df.epi$codes <- sapply(df.epi$Mutation_Type_Code, function(x) strsplit(as.character(x),", "))

isescape <- function(x) "E" %in% x || "LE" %in% x || "IE" %in% x || "NSF" %in% x

df.epi$type <- sapply(df.epi$codes, 
                      function(x) ifelse("TCR" %in% x, "tcr",
                                  ifelse("DHB" %in% x, "hla", 
                                  ifelse(isescape(x), "unknown", 
                                         "non-escape"
                                         ))))

df.epi <- subset(df.epi, type != "non-escape")

df.epi <- data.frame(seq = df.epi$Epitope,
                     aa_from = as.character(sapply(df.epi$Mutation_epitope, 
                                                   function(x) substr(x,1,1))),
                     aa_to = as.character(sapply(df.epi$Mutation_epitope, 
                                                 function(x) substr(x,len(x),len(x)))),
                     pos = sapply(df.epi$Mutation_epitope,
                                  function(x) as.integer(substr(x,2,len(x)-1))),
                     type = as.factor(df.epi$type),
                     elen = sapply(df.epi$Epitope,
                                   function(x) len(x))
                     )

df.epi$pos_adj <- apply(df.epi, 1, function(x) as.numeric(x[4])-as.numeric(x[6]) / 2)

summary(df.epi)
```

As expected, **tcr**-mediate escape mutations are clustered in the central part of antigen (showing a two-peak picture resembling TCR contact positioning plot above), while **hla**-mediate escape mutations are clustered on **N** and **C** termini.

```{r}
ggplot(df.epi, aes(x=pos_adj, fill=type)) +
  annotate("rect", xmin=-Inf, xmax=-1, ymin=0, ymax=Inf, fill="grey") +
  annotate("rect", xmin=3, xmax=Inf, ymin=0, ymax=Inf, fill="grey") +
  geom_density(alpha=0.5) + 
  annotate("text", x=-4, y=0.25, label = "hla(?)") +
  annotate("text", x=-0, y=0.25, label = "tcr(?)") +
  annotate("text", x=4, y=0.25, label = "hla(?)") +
  xlab("Position, relative to epitope center") +
  ylab("") +
  scale_fill_brewer(name = "Escape route", palette = "Set1") +
  scale_x_continuous(limits=c(-6,6), expand=c(0,0)) + 
  scale_y_continuous(limits=c(0,0.26), expand=c(0,0)) +
  theme_bw()
```

Next, we have tested the hypothesis that **tcr**-mediated escape variants are biased in the direction of mutations that substantially change TCR recognition profile (*TRP*, i.e. a column from the contact amino acid matrix described above). We have therefore computed correlation coefficients between *TRPs* of original and substituted amino acids and compared them between **hla**, **tcr** and **other** mutation groups. We have also taken an advantage of positioning data and partitioned the **other** group into **other.tcr** and **other.hla** based on whether the variant was in *[-2, 2]* region in respect to antigen center or not.

```{r}
calc_dist_1 <- function(x , y) cor(contact_mat[,x], contact_mat[,y])

mean_dist <- median(mapply(function(x,y) ifelse(x==y, NA, calc_dist_1(x,y)), 
                         as.character(df.epi$aa_from), 
                         as.character(df.epi$aa_to)), na.rm=T)

calc_dist <- function(row) calc_dist_1(row[2], row[3]) - mean_dist

df.epi$type2 <- factor(apply(df.epi, 1, function(x) ifelse(x[5] == "unknown", 
                                                  ifelse(as.numeric(x[7]) >= -1 & as.numeric(x[7]) <= 3, 
                                                         "tcr(?)", "hla(?)"),
                                                  x[5])), levels=c("hla","hla(?)","tcr(?)","tcr"))

df.epi$dist <- apply(df.epi, 1, calc_dist)

ggplot(df.epi, aes(x = type2, group=type2, y=dist)) +
  geom_violin() + 
  geom_boxplot(aes(fill=type2), width=0.3) + 
  geom_hline(yintercept = 0, linetype="dashed") +
  xlab("Escape route") + 
  ylab("Similarity of CDR3 contact profiles") +
  scale_fill_brewer(guide=F, palette = "Set1") + 
  scale_y_continuous(limits=c(-0.75,0.25)) +
  theme_bw()
```

There was indeed a clear trend showing that **hla**-mediated escape mutations resulted in similar *TRPs*, while **tcr**-mediated mutations had least similar *TRPs*. Statistical details are given below.

```{r}
kruskal.test(dist ~ type2, df.epi)
wilcox.test(subset(df.epi, type2 == "hla(?)")$dist, subset(df.epi, type2 == "tcr(?)")$dist)
wilcox.test(subset(df.epi, type2 == "hla")$dist, subset(df.epi, type2 == "tcr")$dist)
```

The analysis described here can be further extended to immunogenic and non-immunogenic mutated self-peptides from cancer peptidomes once a database of cancer epitopes similar to [NIAID HIV databases](http://www.hiv.lanl.gov/) will become available.